File: android/app/src/main/kotlin/ai/gowtam/oak/services/CalcService.kt
package ai.gowtam.oak.services

import ai.gowtam.oak.networking.Endpoint
import ai.gowtam.oak.networking.OakApiClient
import ai.gowtam.oak.networking.OakError
import ai.gowtam.oak.wire.CalcResult
import ai.gowtam.oak.wire.CalcScenario
import android.util.Log

/**
 * `POST /api/calc` — never-throw (CALC-BR-1). A transport/HTTP/decode fault
 * folds to `null`; in-domain misses ride back as [CalcResult.Error].
 */
interface CalcService {
    suspend fun estimate(scenario: CalcScenario): CalcResult?
}

class LiveCalcService(private val apiClient: OakApiClient) : CalcService {
    override suspend fun estimate(scenario: CalcScenario): CalcResult? {
        val endpoint = Endpoint(
            method = Endpoint.Method.POST,
            path = "/api/calc",
            body = Endpoint.jsonBody(CalcScenario.serializer(), scenario),
            requiresAuth = false,
        )
        return try {
            apiClient.send(endpoint, CalcResult.serializer())
        } catch (e: OakError) {
            Log.e(TAG, "calc estimate unavailable: ${e::class.simpleName}")
            null
        }
    }

    private companion object {
        const val TAG = "Oak.Calc"
    }
}

File: ios/OakApp/Services/CalcService.swift
import Foundation

/// `POST /api/calc` — never throws. A transport / non-2xx / decode miss folds
/// to `nil`; an in-domain miss (`incomplete`, `status_move`, …) is a
/// `.failure` result (CALC-BR-1 / CALC-BR-8).
protocol CalcService: Sendable {
  func estimate(_ scenario: CalcScenario) async -> CalcResult?
}

struct LiveCalcService: CalcService {
  private let apiClient: OakAPIClient

  init(apiClient: OakAPIClient) {
    self.apiClient = apiClient
  }

  func estimate(_ scenario: CalcScenario) async -> CalcResult? {
    do {
      return try await apiClient.send(CalcEndpoints.estimate(scenario), as: CalcResult.self)
    } catch {
      Log.network.error("calc estimate failed")
      return nil
    }
  }
}
